### ISSUE TITLE
Large number of custom emoji causes considerable drop in web interface response time

### ISSUE THREAD

**Issue by codl** [tootsuite/mastodon#6340]
this was observed on an instance running on a low powered server (donphan.social), where `/about` and `/web` take 8 to 10 seconds to load

i've tested this by importing all of mutant standard using [a modified version of Daggertooth's mastomutant.sh](https://gist.github.com/codl/1d4da2717d67b0936c5f35e84f414c11), and by applying a fair amount of cpu limiting (max. 15% of a core) on puma to ensure that responses were cpu-bound, not waiting on postgres or redis

```
$ pgrep -f puma | while read pid; do cpulimit -p $pid -l 15 & done
$ ab -n 10 http://127.0.0.1:3000/about
```

with this test setup, i've observed a jump in median response time from 95ms without any custom emoji, to 11s with all of mutant standard.. that's roughly 100x as much 💦

i'm no profiling expert but fumbling around flamegraphs generated by `rack-mini-profiler` it seems most of that extra time is spent serialising custom emoji and asking paperclip what their url is

**Comment by thekettu** [comment]
I'm running into this on an instance I just deployed Mutant to, though I primarily host my assets off-site, and it seems to be querying the emoji table and then generating JSON on each page load. There must be a better way to cache this so it doesn't take extra time, possibly by moving this to its own statically-generated file?

**Comment by Gargron** [comment]
Tbh I was never expecting any server to use more than 50 emojis (that's the limit in Discord, for example). This is quite extreme. I suppose we should cache the JSON output

**Comment by thekettu** [comment]
@Gargron Yeah, I had been surprised myself, and apparently this is a common request for a lot of server owners to add this. I'm not sure if support for emoji sheets had been talked about in another issue or not, since the one I uploaded is meant to replace the native emoji itself.

(Apparently the emoji-mart component itself is struggling to render 72 some-odd pages of custom emoji as well, and that's kind of expected, but that isn't related to Mastodon itself.)

**Comment by Gargron** [comment]
Fixed by #7047
